tests/rptest/scale_tests/many_partitions_test.py

```diff
@@ -245,142 +176,290 @@ def test_many_partitions(self):
+                any_incomplete = True
+                continue
 
-        self.logger.info(f"Awaiting elections...")
-        wait_until(lambda: self._all_elections_done(topic_names, n_partitions),
-                   timeout_sec=60,
-                   backoff_sec=5)
-        self.logger.info(f"Initial elections done.")
+            assert len(partitions) == p_per_topic
+            for p in partitions:
+                if p.leader == -1:
+                    self.logger.info(
+                        f"partition {tn}/{p.id} has no leader yet")
+                    any_incomplete = True
 
-        for node in self.redpanda.nodes:
-            files = self.redpanda.lsof_node(node)
-            file_count = sum(1 for _ in files)
-            self.logger.info(
-                f"Open files after initial selection on {node.name}: {file_count}"
-            )
+        return not any_incomplete
 
-        # Assume fetches will be 10MB, the franz-go default
-        fetch_mb_per_partition = 10 * 1024 * 1024
+    def _node_leadership_evacuated(self, topic_names: list[str],
+                                   p_per_topic: int, node_id: int):
+        any_incomplete = False
+        for tn in topic_names:
+            partitions = list(self.rpk.describe_topic(tn, tolerant=True))
+            if len(partitions) < p_per_topic:
+                self.logger.info(f"describe omits partitions for topic {tn}")
+                any_incomplete = True
+                continue
 
-        # * Need enough data that if a consumer tried to fetch it all at once
-        # in a single request, it would run out of memory.  OR the amount of
-        # data that would fill a 10MB max_bytes per partition in a fetch, whichever
-        # is lower (avoid writing excessive data for tests with fewer partitions).
-        # * Then apply a factor of two to make sure we have enough data to drive writes
-        # to disk during consumption, not just enough data to hold it all in the batch
-        # cache.
-        write_bytes_per_topic = min(
-            int((self.redpanda.get_node_memory_mb() * 1024 * 1024) / n_topics),
-            fetch_mb_per_partition * n_partitions) * 2
+            assert len(partitions) == p_per_topic
+            for p in partitions:
+                if p.leader == node_id:
+                    self.logger.info(
+                        f"partition {tn}/{p.id} still on node {node_id}")
+                    any_incomplete = True
 
-        if self.scale.release:
-            # Release tests can be much longer running: 10x the amount of
-            # data we fire through the system
-            write_bytes_per_topic *= 10
+        return not any_incomplete
 
-        msg_size = 128 * 1024
-        msg_count_per_topic = int((write_bytes_per_topic / msg_size))
+    def _node_leadership_balanced(self, topic_names: list[str],
+                                  p_per_topic: int):
+        node_leader_counts = defaultdict(int)
```

Comment:
You will like the `Counter` class, I only learned about it recently and it replaced this type of `defaultdict` use (especially nice with typed method because you don't have to separately spell the type of the whole defaultdict type).

Reply:
oh, nice